Add a Best Practices section to the website. NFC (#27334)
emscripten-core/emscripten: site/source/docs/compiling/best_practices.rst

.. _best-practices:

==============
Best Practices
==============

This guide provides recommendations and best practices for compiling C and C++
to the modern web using WebAssembly and Emscripten.

> Following this guide when when reporting bugs an also help speed up diagnosing

Reviewer: ```suggestion
Following this guide when when reporting bugs can also help speed up diagnosing
```

> and fixing of issue since you will be on the well trodden path.

Reviewer: ```suggestion
and fixing issues since you will be on the well-trodden path.
```

In some cases emscripten will guide in the direction of these best practices by

> emitting warning flags, etc, but that is not always possible.

Reviewer: ```suggestion
emitting warnings, etc, but that is not always possible.
```

  main UI thread prevents rendering and freezes the web page. Yield to the event
  loop using ``emscripten_set_main_loop()`` or offload heavy computation to
  background threads using ``-pthread``.
- **Don't rely on raw ``extern "C"`` pointer manipulation** when interacting
  with complex C++ objects across the JavaScript boundary; prefer Embind.
- **Prefer standard compiler flags** over Emscripten-specifc ones where
  possible (for example, prefer ``-pthread`` over ``-sUSE_PTHREADS`` and `-m64`
  over ``-sMEMORY64``).
- **Output ES6 modules** by passing ``-sEXPORT_ES6`` when linking.  This
  avoids polluting the global scope.
- **Use simple comma-separated lists** for list-based settings (for example,
  ``-sEXPORTED_FUNCTIONS=main,malloc`` rather than JSON arrays like
  ``-sEXPORTED_FUNCTIONS=['_main','_malloc']``).
- **Use separate compilation** by compiling ``.cpp`` sources to ``.o`` object
  files before linking rather than combining everything into a single monolithic
  compiler invocation.

Recommended Flags
=================

- ``-sSTRICT``: Opt into strict modern Emscripten behavior, disabling
  deprecated or legacy compatibility features.

> - ``-sEXPORT_ES6``: Output a modern ES6 module (``module.mjs``). This option

Reviewer 2: you mentioned EXPORT_ES6 above. is this section supposed to be a superset of the one above?

Author: Removed.

.. _best-practices:

==============
Best Practices
==============

This guide provides recommendations and best practices for compiling C and C++
to the modern web using WebAssembly and Emscripten.

Following this guide when reporting bugs can also help speed up diagnosing and
fixing issues since you will be on the well-trodden path.

In some cases emscripten will guide in the direction of these best practices by
emitting warnings, etc, but that is not always possible.

General Recommendations
=======================

- **Don't pass settings that are enabled by default.** To keep command lines
  clean and concise, omit redundant options that are already default in modern
  Emscripten (such as ``-sWASM=1``).

> - **Prefer standard compiler flags** over Emscripten-specifc ones where

Reviewer: ```suggestion
- **Prefer standard compiler flags** over Emscripten-specific ones where
```

  possible (for example, prefer ``-pthread`` over ``-sUSE_PTHREADS`` and `-m64`
  over ``-sMEMORY64``).
- **Use simple comma-separated lists** for list-based settings (for example,

>   ``-sEXPORTED_FUNCTIONS=main,malloc`` rather than JSON arrays like

Reviewer 3: You still need the underscores right?

Author: Oh yes, good point :-/

I've been trying to remove that mangling  for almost a decade.. maybe one I will figure out a way.

  ``-sEXPORTED_FUNCTIONS=['_main','_malloc']``).
- **Don't include the "=1" suffix for boolean flags.** For example, write
  ``-sSTRICT`` and ``-sALLOW_MEMORY_GROWTH`` rather than ``-sSTRICT=1`` or
  ``-sALLOW_MEMORY_GROWTH=1``.
- **Use separate compilation** by compiling ``.cpp`` sources to ``.o`` object
  files before linking rather than combining everything into a single monolithic
  compiler invocation.
- **Don't rely on raw ``extern "C"`` pointer manipulation** when interacting

>   with complex C++ objects across the JavaScript boundary; prefer Embind.

Reviewer: This point feels out of place in this list. It is at minimum debatable, I'd say? Raw pointer manipulation is very simple and efficient once you know it.

Generally speaking, I don't think we should recommend embind so generally, given the overhead. It is better that people make the effort to be efficient where it matters, and use embind where speed isn't an issue, but hard to put that into a short bullet point...

Author: Maybe "perfer embind, unless you really know what you doing" although I'm not sure how to say that exactly either.

Reviewer: I think it's best to leave it out. There isn't a simple-enough best practice to recommend.

Author: I think what I'm trying to do here is have newcomers avoid trying do direct pointer manipulation, and present embind as the well-lit path to getting nice integration between native and JS code.

Reviewer: My worry is that people start using embind even for simple things that don't need it, and then later ask "wait why is everything so slow?"

I guess I disagree about embind being the well-lit path. Or maybe it is well-lit but has pitfalls :smile:

Reviewer 2: Could it make sense to have a recommendation based on the type of API being exported to JS? e.g. use Embind if you have a C++ API or you want to do X and Y with your interop, or use something else if it's a C API?

Author: Perhaps we could phrase it as a warning against direct pointer manipulation unless you have a good reason (and good skills) to go that route.

I'm also thinking if recommending using pointer + subarray when passing data around rather than using a lot of calls.   e.g. prefer `getBytesFromWasm(ptr)` over a sequence of `getSingleValueFromWasm()`

Author: I rephrased this point and using `Avoid .. ; prefer ..` rather than being overly prescriptive.

.. _Best-Practices:

==============
Best Practices
==============

This guide provides recommendations and best practices for compiling C and C++
to the modern web using WebAssembly and Emscripten.

Following this guide when reporting bugs can also help speed up diagnosing and
fixing issues since you will be on the well-trodden path.

In some cases, Emscripten will guide you towards these best practices by
emitting warnings, but that is not always possible.

General Recommendations
=======================

- **Don't pass settings that are enabled by default.** To keep command lines
  clean and concise, omit redundant options that are already default in modern
  Emscripten (such as ``-sWASM=1``).
- **Prefer standard compiler flags** over Emscripten-specific ones where
  possible (for example, prefer ``-pthread`` over ``-sUSE_PTHREADS`` and
  ``-m64`` over ``-sMEMORY64``).
- **Use simple comma-separated lists** for list-based settings (for example,
  ``-sEXPORTED_FUNCTIONS=_main,_malloc`` rather than JSON arrays like
  ``-sEXPORTED_FUNCTIONS=['_main','_malloc']``).

> - **Avoid long lists on the command line**; Use the ``@filename`` instead (for

Reviewer: ```suggestion
- **Avoid long lists on the command line**; use a response file instead, with ``@filename`` (for
```

  example, ``-sEXPORTED_FUNCTIONS=@exported_funcs.txt``).
- **Don't include the "=1" suffix for boolean flags.** For example, write
  ``-sSTRICT`` and ``-sALLOW_MEMORY_GROWTH`` rather than ``-sSTRICT=1`` or
  ``-sALLOW_MEMORY_GROWTH=1``.

> - **Use separate compilation** by compiling ``.cpp`` sources to ``.o`` object

Reviewer: I would put all the things before this in a "lint" section. Separate compilation and C++/JS interop feel like substantially different things, choices about the build system and design.

Author: Done

Author: I've called this section "Command-Line Style"  .. I also considered "Command line hygene"

  files before linking rather than combining everything into a single monolithic
  compiler invocation.
- **Avoid direct usage of C/C++ functions from JavaScript**; prefer higher-level
  interfaces such as :js:func:`cwrap` and/or :ref:`embind` that support more

>   than purely numeric types.

Reviewer: I still disagree with this. Direct usage is simply faster, and often works well. How about

* Call directly between C/C++ and JavaScript where possible. When only passing around numeric types this is the most efficient thing to do.
* For more complex things, like passing a `std::vector` between the languages, use embind.

Author: Ok, I replaced completely with a point a void avoiding too many JS/Wasm boundary crossings.

Recommended Flags
=================

- ``-sSTRICT``: Opt into strict modern Emscripten behavior, disabling
  deprecated or legacy compatibility features.
- ``-sEXPORT_ES6``: Output a modern ES6 module (``module.mjs``). This option
  implies ``-sMODULARIZE`` so the generated code will be encapsulated and not
  impact the global namespace.
- ``-sENVIRONMENT=web``: Limit the runtime support to only the environments you
  are targeting. This reduces code size by, for example, omitting Node.js and
  compatibility code.
- ``-Werror -Wall``: Treat warnings as errors to catch C++ bugs and invalid
  compiler settings early.
- ``-O3``, ``-Os``, or ``-Oz``: For release builds, choose ``-O3`` when runtime
  performance is most critical, or ``-Os`` / ``-Oz`` when minimizing binary
  payload size is the priority.
- ``-flto``: Enable Link-Time Optimization (LTO) during both the compilation and
  linking steps of release builds for maximum runtime performance and size

>   reduction.

Reviewer: Should we recommend Closure here?

Author: Done

> 

Reviewer: This section overlaps with our Optimizing Code page. Perhaps link to there? Or perhaps have a section here for optimizations that summarizes that page?

In particular, while most of this list is optimizations, things like STRICT feel more like "lint" (group with "avoid `-sWASM=1`, and EXPORT_ES6 feels more like "design choices" (group with "separate the compile&link stages)

Author: I've created a new "Command line hygiene" section, but lets `-sSTRICT` here under `Recommended flags` because it feels like that is what it is.

that affect code generation (such as ``-flto``, ``-O3``, ``-g``, or
``-pthread``) are passed at **both** compile and link times.

**Compilation step (producing object files):**

.. code-block:: bash

   em++ -O3 -flto -c main.cpp -o main.o $(CXXFLAGS)

**Linking step (producing the ES6 module and WebAssembly binary):**

.. code-block:: bash

   em++ -O3 -flto -sSTRICT -sEXPORT_ES6 --bind main.o -o module.mjs $(LDFLAGS)

Debug vs. Release Profiles
--------------------------

When configuring build profiles, keep compile and link flags consistent within
each configuration:

- **Release Builds:** Use ``-Oz`` or ``-Os`` (or ``-O3`` for CPU-bound tasks)
  combined with ``-flto``.
- **Debug Builds:** Use ``-g`` when compiling and either ``-g``,
  ``-gline-tables-only``, or ``-gsource-map`` when linking. Avoid optimization
  flags (such as ``-O2`` or ``-O3``) or ``-flto`` during debug builds for
  faster compilation and accurate debugging.

Modern Web Workflows and Common Pitfalls
========================================

Interoperability with JavaScript
--------------------------------

When exposing C++ functionality to JavaScript, prefer :ref:`embind` (``--bind``)

> over raw ``extern "C"`` functions. Embind naturally handles C++ classes,

Reviewer: This overlaps with the above text, and I have the same response.

Author: I removed this section.

Asynchronous Code Execution and Main Thread Blocking
----------------------------------------------------

**Don't run long synchronous loops on the browser main thread.** The browser
uses cooperative multitasking; blocking the main UI thread prevents rendering
and freezes the web page.

- Restructure infinite loops to yield to the event loop using
  :c:func:`emscripten_set_main_loop`
  (see :ref:`emscripten-runtime-environment-howto-main-loop`).
- For synchronous-looking C++ code that must pause (or interact with
  asynchronous JavaScript APIs such as ``fetch()`` or Web Promises) without
  refactoring into callbacks, use :ref:`Asyncify <yielding_to_main_loop>`
  (``-sASYNCIFY``) or JavaScript Promise Integration (``-sJSPI``).
- Offload heavy compute or blocking operations to background workers using
  :doc:`multithreading and pthreads <../porting/pthreads>` (``-pthread``).

Virtual Filesystem and I/O
--------------------------

Standard C/C++ file operations (such as ``fopen`` or ``std::ifstream``) operate
on Emscripten's virtual in-memory filesystem (``MEMFS`` by default). See the
:ref:`file-system-overview` for an architectural overview.

> - Do not assume direct access to the host file system.

Reviewer: This isn't a "best practice" so much as a "how to" - ?

Author: Maybe.   I think it is a "what not to do".. which fall under this "Common Pitfalls" section.

Author: I removed this section for now.

- For small temporary files, ``MEMFS`` is sufficient.
- For persistent client-side data storage across browser sessions, use
  asynchronous storage backends such as :ref:`filesystem-api-idbfs` or the

>   :ref:`Filesystem-API`.

Reviewer: Perhaps just link to the main filesystem page for filesystem recommendations? This could also summarize that page, if we want that (if so, perhaps mention WasmFS).

Author: Yes, that makes sense.  Linked to wider docs on FS

Author: I removed this section for now

to the modern web using WebAssembly and Emscripten. Following these practices
ensures clean JavaScript and C++ interoperability, optimal performance,
smaller payload sizes, and future-proof code.

Quick Start and Recommended Flags
=================================

When setting up a new project or modernizing an existing build system for the
web, we recommend starting with the following core compilation and linking
flags:

- ``-sSTRICT``: Opt into strict modern Emscripten behavior, disabling
  deprecated or legacy compatibility features.
- ``-sEXPORT_ES6``: Output a modern ES6 module (``module.mjs``). This option
  implies ``-sMODULARIZE`` so the generated code will be encapsulated and not
  impact the global namespace.
- ``-sENVIRONMENT=web``: Limit the runtime support to only the environments you
  are targeting.  This reduces code size by, for example, omitting Node.js and
  compatibility code.
- ``-Werror -Wall``: Treat warnings as errors to catch C++ bugs and invalid
  compiler settings early.
- ``-Oz`` or ``-Os``: For release builds where payload size is critical, prefer
  ``-Oz`` or ``-Os`` to minimize WebAssembly and JavaScript binary size over

>   ``-O2`` or ``-O3``.

Reviewer: ```suggestion
- Use ``-O3``, ``-Os``, or ``-Oz`` during both compile and link, picking the one that makes the best tradeoff for you in terms of size and speed.
```

Reviewer: I think we should still recommend `-O3`, in other words.

- ``-flto``: Enable Link-Time Optimization (LTO) during both the compilation and
  linking steps of release builds for maximum runtime performance and size
  reduction.
- ``-sALLOW_MEMORY_GROWTH``: Allow the WebAssembly linear memory heap to grow

>   dynamically at runtime, which is essential for many real-world applications.

Reviewer: ```suggestion
```
I don't think this is a best practice. It is necessary for some apps and not for others.

Author: Removed.

- ``--bind``: Enable :ref:`embind` for type-safe, class-based C++ and

>   JavaScript interoperability.

Reviewer: ```suggestion
```
I think we should not recommend embind so broadly. It is great for complex interop, but has a performance penalty.

Dos and Don'ts
==============

The following recommendations highlight best practices and common anti-patterns
when developing with Emscripten:

Dos
---

- **Prefer standard compiler flags** over Emscripten-specifc ones where
  possible (for example, prefer ``-pthread`` over ``-sUSE_PTHREADS`` and `-m64`
  over ``-sMEMORY64``).
- **Output modern ES6 modules** by passing ``-sEXPORT_ES6`` when linking.  This
  avoids polluting the global scope.
- **Use simple comma-separated lists** for list-based settings (for example,
  ``-sEXPORTED_FUNCTIONS=main,malloc`` rather than JSON arrays like
  ``-sEXPORTED_FUNCTIONS=['_main','_malloc']``).
- **Use separate compilation** by compiling ``.cpp`` sources to ``.o`` object
  files before linking rather than combining everything into a single monolithic
  compiler invocation.
- **Use Embind** (``--bind``) to pass complex data types, classes, vectors, and

>   strings cleanly across the JavaScript and C++ boundary.

Reviewer: ```suggestion
```

Don'ts
------

- **Don't pass settings that are enabled by default.** To keep command lines
  clean and concise, omit redundant options that are already default in modern
  Emscripten (such as ``-sWASM=1``).
- **Don't include the "=1" suffix for boolean flags.** For example, write
  ``-sSTRICT`` and ``-sALLOW_MEMORY_GROWTH`` rather than ``-sSTRICT=1`` or

>   ``-sALLOW_MEMORY_GROWTH=1``.

Reviewer: These two are basically minor nits. Maybe worth putting in a "style" section instead?

Author: I restructures to these come update general recommendations.    I completely removed the Do's And Don't idea.

> - **Don't run long synchronous loops on the browser main thread.** Blocking the

Reviewer 2: is there a separate page on this topic that we can link to here? likewise with filesystems, these are both bigger topics that are as much about the application architecture as the flags and configuration.

Reviewer: +1, "don't include unneeded flags or =1" is a minor style nit for developers, but not blocking the main thread is at minimum a large quality issue for users. It feels strange to have both in the same list.

  main UI thread prevents rendering and freezes the web page. Yield to the event
  loop using ``emscripten_set_main_loop()`` or offload heavy computation to
  background threads using ``-pthread``.
- **Don't rely on raw ``extern "C"`` pointer manipulation** when interacting
  with complex C++ objects across the JavaScript boundary; prefer Embind.
- **Don't assume direct synchronous access to host or browser filesystems.**
  Standard I/O functions operate on an in-memory virtual filesystem (``MEMFS``).
  For persistent client-side data, use asynchronous options like ``IDBFS``.

Compiler Warnings and Automated Guidance

> ========================================

Reviewer: Not sure about this section. The warnings we provide should already be self-explanatory?

Author: Removed

detailed in this guide rather than relying solely on warnings in the compiler
output.

Separate Compilation Workflows
==============================

For non-trivial projects, always separate the compilation step (compiling source
files to object files) from the linking step (combining object files into the
final WebAssembly and JavaScript outputs). This enables incremental builds and
matches standard C/C++ development practices.

When using separate compilation, ensure that optimization flags and settings
that affect code generation (such as ``-flto``, ``-O3``, ``-g``, or
``-pthread``) are passed at **both** compile and link times.

**Compilation step (producing object files):**

.. code-block:: bash

   em++ -O3 -flto -c main.cpp -o main.o $(CXXFLAGS)

**Linking step (producing the ES6 module and WebAssembly binary):**

.. code-block:: bash

   em++ -O3 -flto -sSTRICT -sEXPORT_ES6 --bind main.o -o module.mjs $(LDFLAGS)

Debug vs. Release Profiles
--------------------------

When configuring build profiles, keep compile and link flags consistent within
each configuration:

- **Release Builds:** Use ``-Oz`` or ``-Os`` (or ``-O3`` for CPU-bound tasks)

>   combined with ``-flto``.

Reviewer: Same comment as above

- **Debug Builds:** Use ``-g`` to preserve debug symbols and DWARF debugging
  information. Avoid ``-O`` or ``-flto`` flags during debug builds for faster

>   compilation and accurate step-through debugging.

Reviewer: Can mention source maps?

Modern Web Workflows and Common Pitfalls
========================================

Interoperability with JavaScript
--------------------------------

When exposing C++ functionality to JavaScript, prefer :ref:`embind` (``--bind``)
over raw ``extern "C"`` functions. Embind naturally handles C++ classes,
overloaded functions, smart pointers, ``std::string``, and ``std::vector``
without requiring manual memory conversions or unsafe casting in JavaScript.

Asynchronous Code Execution
---------------------------

If your C++ code needs to interact with asynchronous JavaScript APIs (such as
``fetch()`` or Web Promises), avoid blocking or polling. Instead, use
:ref:`asyncify section` (``-sASYNCIFY``) or JavaScript Promise Integration
(``-sJSPI``) to pause and resume the C++ call stack when waiting for
asynchronous operations to complete.

Virtual Filesystem and I/O
--------------------------

Standard C/C++ file operations (such as ``fopen`` or ``std::ifstream``) operate
on Emscripten's virtual in-memory filesystem (``MEMFS`` by default).

> - Do not assume direct access to the host or browser file system.

Reviewer: This section confuses me. "Do not assume" - is this saying not to use NODERAWFS?

- For small temporary files, ``MEMFS`` is sufficient.
- For persistent client-side data storage across browser sessions, use asynchronous
  storage backends such as ``IDBFS`` or the Emscripten Filesystem API.

Networking and Sockets
----------------------

Standard POSIX and BSD sockets (such as raw TCP or UDP sockets) cannot connect
directly across the web due to browser security sandboxing. Instead of relying
on standard socket libraries, use:

- WebSockets for bidirectional, real-time messaging.
- WebRTC for peer-to-peer data channels and UDP-like transport.

> - The Emscripten Fetch API for HTTP and HTTPS requests.

Reviewer: I think this section is handled in the separate networking page?
